aten/src/ATen/native/cuda/SegmentReduce.cu

```diff
@@ -61,20 +93,36 @@ Tensor _segment_reduce_cuda_kernel(
         auto* data_data_ptr = data.data_ptr<scalar_t>();
         auto* output_data_ptr = output.data_ptr<scalar_t>();
 
-        CustomMax max_op{};
-        scalar_t initial_value = initial.has_value()
-            ? initial.value().to<scalar_t>()
-            : std::numeric_limits<scalar_t>::lowest();
-        CUB_WRAPPER(
-            cub::DeviceSegmentedReduce::Reduce,
-            data_data_ptr,
-            output_data_ptr,
-            segment_count,
-            offsets_data_ptr,
-            offsets_data_ptr + 1,
-            max_op,
-            initial_value,
-            at::cuda::getCurrentCUDAStream());
+        if (reduction == SegmentReductionType::MAX) {
+          CustomMax max_op{};
+          scalar_t initial_value = initial.has_value()
+              ? initial.value().to<scalar_t>()
+              : std::numeric_limits<scalar_t>::lowest();
+          CUB_WRAPPER(
+              cub::DeviceSegmentedReduce::Reduce,
+              data_data_ptr,
+              output_data_ptr,
+              segment_count,
+              offsets_data_ptr,
+              offsets_data_ptr + 1,
+              max_op,
+              initial_value,
+              at::cuda::getCurrentCUDAStream());
+        } else if (reduction == SegmentReductionType::MEAN) {
+          scalar_t initial_value =
```

Comment:
It would be better to use cub here also, and divide result by lengths. You can put special initial value handling in the kernel dividing by lengths if you want.

Reply:
sounds good!